### PR TITLE
[coap] optimize call to GetMatchedResponseCopy

### PR DESCRIPTION
**Pull request by simonlingoogle**
Method `ResponsesQueue::EnqueueResponse` was using `GetMatchedResponseCopy` to check if matched response already exists in the cache. The code might clone message only to release it right away, which is not efficient. 

The commits ...
1. Add a new method `ResponsesQueue::FindMatchedResponse` which try to find and return the matched response in the queue, without cloning the message
2. Optimize `EnqueueResponse` by using `FindMatchedResponse`
3. Re-implement `GetMatchedResponseCopy` using `FindMatchedResponse` to  to reduce code redundancy.

### CODE REVIEW

File: src/core/coap/coap.hpp [hunk 1]
```diff
@@ -396,6 +396,19 @@ class ResponsesQueue
...
+     *
+     * @param[in]  aRequest      The CoAP message containing Message ID.
+     * @param[in]  aMessageInfo  The message info containing source endpoint address and port.
+     * @param[out] aResponse     A pointer to the cached CoAP response matching given arguments, or NULL.
+     *
+     * @retval true   Matching response found.
+     * @retval false  Matching response not found.
+     *
+     */
+    bool FindMatchedResponse(const Message &aRequest, const Ip6::MessageInfo &aMessageInfo, Message **aResponse);
```
Comment (bukepo) [inline src/core/coap/coap.hpp]: -> `Message *FindMatchedResponse(const Message &aRequest, const Ip6::MessageInfo &aMessageInfo) const;`?

Comment (bukepo) [reply src/core/coap/coap.hpp]: Is this solved?

Comment (simonlingoogle) [reply src/core/coap/coap.hpp]: yes

Comment (bukepo) [reply src/core/coap/coap.hpp]: I didn't see there's any change about the function prototype.

Comment (simonlingoogle) [reply src/core/coap/coap.hpp]: Sorry, I thought you just meant to add const. 
Prototype modified now.

File: src/core/coap/coap.cpp [hunk 2]
```diff
@@ -711,14 +711,33 @@ otError ResponsesQueue::GetMatchedResponseCopy(const Message &         aRequest,
...
                                                Message **              aResponse)
 {
-    otError                error = OT_ERROR_NOT_FOUND;
-    Message *              message;
-    EnqueuedResponseHeader enqueuedResponseHeader;
-    Ip6::MessageInfo       messageInfo;
+    otError  error = OT_ERROR_NONE;
+    Message *cacheResponse;
+
+    if ((cacheResponse = FindMatchedResponse(aRequest, aMessageInfo)) == NULL)
```
Comment (bukepo) [inline src/core/coap/coap.cpp]: ->
```
cacheResponse = FindMatchedResponse(aRequest, aMessageInfo);
VerifyOrExit(cacheResponse != NULL, error = OT_ERROR_NOT_FOUND);
```

File: src/core/coap/coap.cpp [hunk 3]
```diff
@@ -711,14 +711,33 @@ otError ResponsesQueue::GetMatchedResponseCopy(const Message &         aRequest,
...
+    {
+        ExitNow(error = OT_ERROR_NOT_FOUND);
+    }
+
+    if ((*aResponse = cacheResponse->Clone(cacheResponse->GetLength() - sizeof(EnqueuedResponseHeader))) == NULL)
```
Comment (bukepo) [inline src/core/coap/coap.cpp]: Use the same trick as above to improve code coverage.

File: src/core/coap/coap.hpp [hunk 4]
```diff
@@ -396,6 +396,17 @@ class ResponsesQueue
...
+    /**
+     * Check if CoAP response exists in the cache that matches given Message ID and source endpoint.
+     *
+     * @param[in]  aRequest      The CoAP message containing Message ID.
+     * @param[in]  aMessageInfo  The message info containing source endpoint address and port.
+     *
+     * @retval Matching cached response or NULL if not found
+     *
+     */
+    Message *FindMatchedResponse(const Message &aRequest, const Ip6::MessageInfo &aMessageInfo) const;
```
Comment (bukepo) [inline src/core/coap/coap.hpp]: Should this be `const Message*`?

File: src/core/coap/coap.hpp [hunk 5]
```diff
@@ -397,6 +397,17 @@ class ResponsesQueue
         kMaxCachedResponses = OPENTHREAD_CONFIG_COAP_SERVER_MAX_CACHED_RESPONSES,
     };
 
+    /**
+     * Check if CoAP response exists in the cache that matches given Message ID and source endpoint.
+     *
+     * @param[in]  aRequest      The CoAP message containing Message ID.
+     * @param[in]  aMessageInfo  The message info containing source endpoint address and port.
+     *
+     * @retval Matching cached response or NULL if not found
```
Comment (bukepo) [inline src/core/coap/coap.hpp]: `@retval` -> `@returns`?

Comment (simonlingoogle) [reply src/core/coap/coap.hpp]: thanks

**Pull request merged**